src/item_factory.cpp

```diff
@@ -1994,16 +1994,61 @@ void armor_portion_data::deserialize( const JsonObject &jo )
     } else {
         optional( jo, false, "encumbrance", encumber, 0 );
     }
+    optional( jo, false, "material_thickness", thickness, 0.0f );
+    optional( jo, false, "environmental_protection", env_resist, 0 );
+    optional( jo, false, "environmental_protection_with_filter", env_resist_w_filter, 0 );
+
+    // TODO: Make mandatory - once we remove the old loading below
+    optional( jo, false, "material", materials );
+}
+
+template<typename T>
+static void apply_optional( T &value, const cata::optional<T> &applied )
+{
+    if( applied ) {
+        value = *applied;
+    }
 }
 
 void islot_armor::load( const JsonObject &jo )
 {
     optional( jo, was_loaded, "armor", data );
+
+    cata::optional<float> thickness;
+    cata::optional<int> env_resist;
+    cata::optional<int> env_resist_w_filter;
+    cata::optional<std::vector<material_id>> materials;
+    cata::optional<body_part_set> covers;
+
+    bool all_data_have_material = true;
+    if( jo.has_member( "armor" ) ) {
+        for( const JsonObject &jo : jo.get_array( "armor" ) ) {
+            jo.allow_omitted_members();
+            if( !jo.has_member( "material" ) ) {
+                all_data_have_material = false;
+            }
```

Comment:
This is too close to the code that crashed in the MXE cross compile. It's safer to avoid shadowing here. Also deal with the check error at the same time.
```suggestion
    if( jo.has_member( "armor" ) ) {
        for( JsonObject armor_json : jo.get_array( "armor" ) ) {
            armor_json.allow_omitted_members();
            if( !armor_json.has_member( "material" ) ) {
                all_data_have_material = false;
            }
```

Reply:
Also we shouldn't be shadowing variables at all, it's a terrible misfeature that the language allows it.

Reply:
Should we start adding stricter `-Wshadow` flags to our various settings?